src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs

```diff
@@ -0,0 +1,359 @@
+            }
+
+            var shellSettings = shellContext.Settings;
+
+            if (shellSettings.State != TenantState.Disabled)
+            {
+                _notifier.Error(H["You can only enable a Disabled shell."]);
+            }
+
+            shellSettings.State = TenantState.Running;
+            _orchardHost.UpdateShellSettings(shellSettings);
+
+            return RedirectToAction(nameof(Index));
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Reload(string id)
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.ManageTenants))
+            {
+                return Unauthorized();
+            }
+
+            if (_currentShellSettings.Name != ShellHelper.DefaultShellName)
+            {
+                return Unauthorized();
+            }
+
+            var shellContext = _orchardHost
+                .ListShellContexts()
+                .OrderBy(x => x.Settings.Name)
+                .Where(x => String.Equals(x.Settings.Name, id, StringComparison.OrdinalIgnoreCase))
+                .FirstOrDefault();
+
+            if (shellContext == null)
+            {
+                return NotFound();
+            }
+
+            var shellSettings = shellContext.Settings;
+            _orchardHost.ReloadShellContext(shellSettings);
+
+            return RedirectToAction(nameof(Index));
+        }
+        private void ValidateViewModel(EditTenantViewModel model, bool newTenant)
+        {
+            if (String.IsNullOrWhiteSpace(model.Name))
+            {
+                ModelState.AddModelError(nameof(EditTenantViewModel.Name), S["The tenant name is mandatory."]);
+            }
+
+            var allShells = _orchardHost
+                .ListShellContexts()
+                .OrderBy(x => x.Settings.Name);
+
+            if (newTenant && allShells.Any(tenant => String.Equals(tenant.Settings.Name, model.Name, StringComparison.OrdinalIgnoreCase)))
+            {
+                ModelState.AddModelError(nameof(EditTenantViewModel.Name), S["A tenant with the same name already exists.", model.Name]);
+            }
+
+            if (!String.IsNullOrEmpty(model.Name) && !Regex.IsMatch(model.Name, @"^\w+$"))
+            {
+                ModelState.AddModelError(nameof(EditTenantViewModel.Name), S["Invalid tenant name. Must contain characters only and no spaces."]);
+            }
+
+            if (!String.Equals(model.Name, ShellHelper.DefaultShellName, StringComparison.OrdinalIgnoreCase) && string.IsNullOrWhiteSpace(model.RequestUrlHost) && string.IsNullOrWhiteSpace(model.RequestUrlPrefix))
+            {
+                ModelState.AddModelError(nameof(EditTenantViewModel.RequestUrlPrefix), S["Host and Url Prefix can not be empty at the same time."]);
+            }
+        }
+
+        private IEnumerable<ShellContext> GetShells()
+        {
+            return _orchardHost.ListShellContexts().OrderBy(x => x.Settings.Name);
```

Comment:
Listing tenants in the admin is not a very hot path ... kind of doesn't matter in this case.